### PR TITLE
OCPCLOUD-1711: Resources and logs gatherer for failed tests

### PR DESCRIPTION
**Pull request by lobziik**
This adds a gatherer helper for dumping resources and pod logs for failed tests.

This PR contains:

- tiny wrapper around `oc` CLI, which allows to leverage oc binary for retrieving resources and pod logs
- `StateGatherer` helper, which provides shortcuts for dump MAPI related resources and pod logs
- enables gathering for failed tests across the suite
- (also, .ci-operator.yaml was added, just by the way )

Notes:
* This thing relies on `oc` presence within PATH. In CI this requires `oc` to be injected https://docs.ci.openshift.org/docs/architecture/step-registry/#injecting-the-oc-cli

### CODE REVIEW

File: pkg/framework/gatherer/oc-cli.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CLI provides function to call the OpenShift CLI, which is using to simplify state gathering during tests
+type CLI struct {
```
Comment (JoelSpeed) [inline pkg/framework/gatherer/oc-cli.go:20]: Is this copied from origin? Wonder if we can just import it? Or maybe move it to a library that's easier to import?

File: pkg/providers/aws.go [hunk 2]
```diff
@@ -28,19 +29,30 @@ var _ = Describe("[Feature:Machines] [AWS] MetadataServiceOptions", func() {
...
 
 	AfterEach(func() {
 		Expect(framework.DeleteMachineSets(client, toDelete...)).To(Succeed())
 		toDelete = make([]*machinev1.MachineSet, 0, 3)
 
 		framework.WaitForMachineSetsDeleted(client, toDelete...)
+
+		testDescription := CurrentGinkgoTestDescription()
+		if testDescription.Failed == true {
+			Expect(gatherer.WithTestDescription(testDescription).GatherAll()).To(Succeed())
```
Comment (JoelSpeed) [inline pkg/providers/aws.go:54]: Will the gatherer output go into a folder per test?

File: pkg/framework/gatherer/data-gatherer.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	mapiMachines            = "machines.machine.openshift.io"
+	mapiMachinesets         = "machinesets.machine.openshift.io"
+	controlPlaneMachinesets = "controlplanemachinesets.machine.openshift.io"
+
+	machineApproverNamespace = "openshift-cluster-machine-approver"
+)
+
+var namespacedResourceToGather = [...]string{mapiMachinesets, mapiMachines, controlPlaneMachinesets}
+
+var clusterResourceToGather = [...]string{"nodes", "clusterautoscaler", "machineautoscaler"}
```
[the remark this answers is not in the document]
Comment (lobziik) [reply pkg/framework/gatherer/data-gatherer.go:22]: Its a fixed-size array, `[...]` tells the compiler to figure out a number of elements at compile time.

File: pkg/providers/aws.go (hunk 2 above)
Comment (lobziik) [reply pkg/providers/aws.go:54]: Yup.

File: pkg/framework/gatherer/oc-cli.go (hunk 1 above)
Comment (lobziik) [reply pkg/framework/gatherer/oc-cli.go:20]: I tried to import this thing from origin, but this brings A LOT of new dependencies and plenty of complications with wrangling this.

>Or maybe move it to a library that's easier to import?

For now, this thing doesn't tie much to other code around and is more-less independent. We could move this file out to a separate module if will we need to, but I don't think that it's necessary right now.

File: pkg/framework/gatherer/data-gatherer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply pkg/framework/gatherer/data-gatherer.go:22]: You beat me to it, IMO, prefer simple over clever, this is a rare syntax and I won't be the only experienced Go engineer who looks at that and goes, huh...

File: pkg/framework/gatherer/oc-cli.go (hunk 1 above)
Comment (JoelSpeed) [reply pkg/framework/gatherer/oc-cli.go:20]: A comment mentioning where this code came from would be useful I think

File: pkg/framework/gatherer/data-gatherer.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		}
+	}
+	return nil
+}
+
+func (sg *StateGatherer) GatherPodLogs() {
+	logsSubPath := sg.getSubPath("logs")
+	sg.CLI.WithSubPath(logsSubPath).DumpPodLogsSinceTime(sg.ctx, sg.sinceTime)
+	sg.CLI.WithSubPath(logsSubPath).WithNamespace(machineApproverNamespace).DumpPodLogsSinceTime(sg.ctx, sg.sinceTime)
+}
```
[the remark this answers is not in the document]
Comment (lobziik) [reply pkg/framework/gatherer/data-gatherer.go:74]: No at the moment. I need to figure out how to collect errors properly and do not break the collection process. For now, this function spits log message if smth goes wrong.

File: pkg/framework/framework.go [hunk 5]
```diff
@@ -183,3 +187,55 @@ func WaitForEvent(c runtimeclient.Client, kind, name, reason string) error {
 		return false, nil
 	})
 }
+
+func NewCLI() (*gatherer.CLI, error) {
```
[the remark this answers is not in the document]
Comment (lobziik) [reply pkg/framework/framework.go:194]: Added comments

File: pkg/framework/gatherer/data-gatherer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply pkg/framework/gatherer/data-gatherer.go:85]: Maybe add a TODO comment?

File: pkg/infra/infra.go [hunk 6]
```diff
@@ -145,8 +145,13 @@ var _ = Describe("[Feature:Machines] Managed cluster should", func() {
 	var machineSet *machinev1.MachineSet
 	var machineSetParams framework.MachineSetParams
 
+	gatherer, err := framework.NewGatherer()
+	Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/infra/infra.go:149]: You might have issues with calling an expectation outside of a `BeforeEach`/`It`, you probably want to `var gatherer framework.Gatherer` and then run this bit of the code in the `BeforeEach`

Comment (lobziik) [reply pkg/infra/infra.go]: Yeah, agree. Done.

**Comment by lobziik** [comment]
Gathered state from the failed test might be found here: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-api-actuator-pkg/253/pull-ci-openshift-cluster-api-actuator-pkg-master-e2e-aws-operator/1590328052108234752/artifacts/e2e-aws-operator/test/artifacts/machine-api-e2e-suite/

File: pkg/framework/gatherer/data-gatherer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lobziik) [reply pkg/framework/gatherer/data-gatherer.go]: added

**Comment by JoelSpeed** [comment]
Results look awesome!
/approve

**Comment by JoelSpeed** [comment]
/lgtm
/retest-required

There's a few todos but they can be sorted in follow up PRs, this is great Denis, thanks! :)

**Review by damdo — COMMENTED** [review]
Very nice! Thanks for this
/lgtm

**Pull request merged**